### ISSUE TITLE
DoNotSaveData saves data

### ISSUE THREAD

**Issue by rdebath** [Microsoft/AL#1905]
For 2018cu3 on-premise

The `Uninstall-NAVApp` command will remove all table rows when given the `-DoNotSaveData` flag, however, extension metadata including the data dictionary of the additional database columns is not removed by this nor by `Unpublish-NAVApp`. This means that if a breaking change is made (eg: change a column from Int to Decimal or just rename a column) the republish of the extension will be rejected.

Workarounds include one of:
* Delete the test database.
* Change the uuid of the extension
* Delete all references to the extension uuid from the SQL level (ie: extra tables and rows in `$ndo$navappschemasnapshot` and `$ndo$navappschematracking`).

Note: Bad Things™ also happen if the base table is renamed!

**Comment by HunterFeltman-MSFT** [comment]
I think you are looking for the `Sync-NAVApp` cmdlet [(Doc).](https://docs.microsoft.com/en-us/powershell/module/microsoft.dynamics.nav.apps.management/sync-navapp?view=dynamicsnav-ps-2018)
Also see issues: #1499, #1198, #905, & #1595

**Comment by rdebath** [comment]
It didn't seem to do anything useful:

```
PS C:\temp> Sync-NAVApp -ServerInstance  DemoDatabaseNAV110 -name 'TestExtension' -Mode Clean -Force
WARNING: The extension 'TestExtension' is not synchronized.
PS C:\temp> Sync-NAVApp -ServerInstance  DemoDatabaseNAV110 -name 'TestExtension' -Mode Add -Force
WARNING: Cannot synchronize the extension TestExtension because it is already synchronized.
```

But with your prompting I tried again; it appears that it does work IF you are psychic enough to know you need to delete the data before vs-code tells you that it can't update.

It _might_ also work if I actually committed the previously broken version and can go back to it. I expect it'll be easier to just delete the trash manually.

So no I don't think it's working correctly.

**Comment by rdebath** [comment]
Okay, even more time spent and eventually I tried something that's completely forbidden by the documentation. I tried running `Sync-NavApp -Mode Clean -Force` against an extension (V2) that is NOT installed and NOT published.

Kudos to the developer, it seems to clean up everything, even damaged data.
Contempt to the documentation writer who complete messed it up!

In addition, it seems that the state I had that test extension in is some sort of exception as usually the "Clean" does work even after vs-code has been trying to update an extension.

I would suggest that, for V2 extensions, `-DoNotSaveData` is specifically documented as not working as expected (that is it doesn't delete everything). Then that `Sync-NavApp -Mode Clean` is clearly documented as working both **post-uninstall** and **post-unpublish**. 

Also the `-DoNotLoadData` flag for `Install-NavApp` is badly broken, at the very least it should be documented as "Do Not Use on V2 extensions". But a reasonable fix would be to do a full clean of that tenant before installing even though this isn't exactly what it says on the tin.

**Comment by HunterFeltman-MSFT** [comment]
What documentation are you referring to that states that this action is forbidden when the extension is neither install nor published? (In fact the extension must always be uninstalled first elsewise we would be deleting data that is currently in use!) That scenario is definitely supported for the 'clean' mode and so such documentation should be updated to reflect this.

As to the flags `-DoNotSaveData` & `-DoNotLoadData`, the expected result of their usage is only that they purge the data stored in the tables & table extensions that come from the target extension. They do not deal with truncating or updating physical schema elements. That, as you have discovered, is the job of `Sync-NAVApp `. 

You mentioned in your inital post that you were directly truncating some of the system DB records, such as $ndo$navappschematracking, which are used by the platform to determine the answer to questions like 'what extensions are in sync' and 'does this new extension that we are trying to sync create a breaking schema change', direct modification to these tables is definitely not recommended as it could put the platform in a bad state.

I am still interesting in the 'exception' case you were able to put your extension into that prevented any sync action from happening. Do you still have solid repro steps to get into this case that does not involve modifying the DB directly?

**Comment by rdebath** [comment]
HunterFeltman-MSFT's URL:  https://docs.microsoft.com/en-us/powershell/module/microsoft.dynamics.nav.apps.management/sync-navapp?view=dynamicsnav-ps-2018

Description: "Use the Sync-NAVApp cmdlet to synchronize the schema changes from an app to a tenant database."

This is simply saying to me that once you've "loaded the objects" you "do the sync" just like normal objects (but the Install-NavApp says this is done automatically).

The clean mode does have a line at the end saying "This mode should only be used as a final schema cleanup step when the NAV App and all versions are intended to be removed."

The English is a little strained, but notice the word "intended", clearly saying you do this BEFORE you "remove" the extension. Now, that doesn't sound unreasonable, but, "remove" could mean "uninstall" or "unpublish" so I tried them both on a working extension and doing it while installed gave a reasonable error message. Doing it before unpublish gave the two "nothing to do" messages (on the broken one). 

Once I realised that `Sync-NAVApp` worked when the extension wasn't even published I did try it on some examples where I intentionally broke things at the SQL level; it had no problems.

I'm not that worried about the `-DoNotSaveData`, as a user I would read that as it removes all traces of the extension, but a user isn't likely to notice or care about the (usually) minor leftovers. As a developer I realise that the "data" and the "metadata" are somewhat different, though it feels a little untidy.

OTOH, `-DoNotLoadData` is broken, it successfully installs an extension and allows it to be used without doing a schema sync. The extension was up and running but the SQL table had the previous structure. If the structure isn't "compatible" at the SQL level you get SQL errors.



As for the exception, unfortunately, I wasn't able to duplicate it, though it was before I touched the database at the SQL level. I had been experimenting with two different but _very_ similar projects, seeing what I got with various sorts of collisions between the extensions and other breaking changes. 

WRT the manual clear up, before I cleared the SQL data I removed the extension (uninstall and unpublish), which was successful, Then I removed everything left to do with the UUID in question. The database did not have any indication of leftover data and was working fine thereafter.

**Comment by HunterFeltman-MSFT** [comment]
Thanks for the walkthrough, I can see how you might come to those conclusions based on the wording. I'll talk with the doc teams to see what we can do about restructuring to make things more clear (for both sync & install)

I'll also take a look into the 'broken' `-DoNotLoadData` issue and try to see what's up there.

**Comment by HunterFeltman-MSFT** [comment]
@rdebath I was finally able to get back to this issue, but I'm afraid I haven't been able to reproduce the issue you noticed with the `-DoNotLoadData` flag.

At not point was I able to 'bypass' the Sync-NAVApp restriction on extension by using this flag. I was always required to have synchronized the extension to at least the version specified in the package that was being installed.

The only case where it 'appeared' to have worked was when I created a similar extension whereby I used the same AppId + version in the app.json and synchronized that. In this case, the system thinks that both of the extensions are the 'same' (since the AppId is used to identify a 'family' of extensions), and it is 'tricked' into letting the install happen.

I noticed that you mentioned you were dealing with **_very_** similar projects, is it perhaps the case where the AppId's of these projects were the same?

If not, could you let me know/provide specific examples where you were able to do this bypass?

**Comment by HunterFeltman-MSFT** [comment]
I think that confusion about the reason for not removing schema elements is due to another 'holdover' in the documentation around install/uninstall from V1, which I have notified the doc team about. 

For V2 extensions, the install/uninstall _cmdlets_ do **not** handle synchronization for you, this is handled exclusively by Sync-NAVApp.

This is to support the '[Shared Schema](https://docs.microsoft.com/en-us/powershell/module/microsoft.dynamics.nav.management/sync-navtenantdatabase?view=dynamicsnav-ps-2018)' database mode whereby multiple tenants can be housed within the same database. These tenants all share the same 'schema' and have their data (logically isolated), stored in the same tables. This mode is currently, to my knowledge, only really utilized in Dynamics 365/BC.

Still, by allowing multiple tenants to exist in the same database, we have to account for multiple extensions or multiple versions of a extension all coexisting in that database. Additionally, we have to support multiple tenants using those extensions and installing/uninstalling them 'at whim'. To accomplish that, we implemented the idea of companion tables that you see today, which isolate the schema of a given extension from its associated 'base' table, however, the data inside of these tables is still 'shared' by tenants.

As such, we can no longer roll schema 'off' during uninstall as a given tenant may not be the only one 'using' it. (And we wouldn't want to send some else's data to the abyss)

I agree with your suggesting/direction of utilizing the `Sync-NAVApp` cmdlet's 'clean' mode during development as it ensures a consistent 'fresh' state when iterating through early extension pseudo-versions.
